Use consistent names for internal `nvcc` files

trxcllnt opened the pull request:
This PR extracts the fixes changes from https://github.com/mozilla/sccache/pull/2356 per @drahnr's request.

This PR branched from/is a follow-up to https://github.com/mozilla/sccache/pull/2382. See the diff of the two branches [here](https://github.com/trxcllnt/sccache/compare/fea/enhance-system-tests...trxcllnt:sccache:fix/consistent-nvcc-internal-file-names?expand=1).

The names for the internal files depend on the compilation flag and device architectures. `nvcc` generates a different name for the `.cpp1.ii`, `.cudafe1.c`, `.cudafe1.stub.c`, `.cudafe1.gpu`, `.ptx`, and `.cubin` files when the compile flag is `-ptx`, `-cubin`, or `-c`, and also on whether there's one vs. many `-gencode` arguments. Additionally, it will either include or omit the `--gen_module_id_file` flag from the `cicc` invocation based on whether the compile flag is `-ptx`, `-cubin`, or `-c`.

Some examples:
```shell
# compile flag: -ptx, single arch
$ nvcc -x cu -ptx x.cu -o x.cu.o -gencode=arch=compute_60,code=[compute_60,sm_60] --dryrun --keep --keep-dir /x 2>&1 | grep -P '(cpp1\.ii|\.ptx|\.cubin)'
#$ gcc -E ... "x.cu" -o "/x/x.cpp1.ii" 
#$ "$CICC_PATH/cicc" ... --gen_module_id_file --module_id_file_name "/x/x.module_id"  "/x/x.cpp1.ii" -o "x.cu.o"

# compile flag: -cubin, single arch
$ nvcc -x cu -cubin x.cu -o x.cu.o -gencode=arch=compute_60,code=[sm_60] --dryrun --keep --keep-dir /x 2>&1 | grep -P '(cpp1\.ii|\.ptx|\.cubin)'
#$ gcc -E ... "x.cu" -o "/x/x.cpp1.ii" 
#$ "$CICC_PATH/cicc" ... --gen_module_id_file --module_id_file_name "/x/x.module_id" --gen_c_file_name "/x/x.cudafe1.c" --stub_file_name "/x/x.cudafe1.stub.c" --gen_device_file_name "/x/x.cudafe1.gpu"  "/x/x.cpp1.ii" -o "/x/x.ptx"
#$ ptxas -arch=sm_60 -m64  "/x/x.ptx"  -o "x.cu.o"

# compile flag: -c, single arch
$ nvcc -x cu -c x.cu -o x.cu.o -gencode=arch=compute_60,code=[compute_60,sm_60] --dryrun --keep --keep-dir /x 2>&1 | grep -P '(cpp1\.ii|\.ptx|\.cubin)'
#$ gcc -E ... "x.cu" -o "/x/x.cpp1.ii" 
#$ "$CICC_PATH/cicc" ... --module_id_file_name "/x/x.module_id" --gen_c_file_name "/x/x.cudafe1.c" --stub_file_name "/x/x.cudafe1.stub.c" --gen_device_file_name "/x/x.cudafe1.gpu"  "/x/x.cpp1.ii" -o "/x/x.ptx"
#$ ptxas -arch=sm_60 -m64  "/x/x.ptx"  -o "/x/x.sm_60.cubin" 

# compile flag: -c, multiple archs
$ nvcc -x cu -c x.cu -o x.cu.o -gencode=arch=compute_60,code=[sm_60] -gencode=arch=compute_70,code=[compute_70,sm_70] --dryrun --keep --keep-dir /x 2>&1 | grep -P '(cpp1\.ii|\.ptx|\.cubin)'
#$ gcc -E ... "x.cu" -o "/x/x.compute_60.cpp1.ii" 
#$ "$CICC_PATH/cicc" ... --module_id_file_name "/x/x.module_id" --gen_c_file_name "/x/x.compute_60.cudafe1.c" --stub_file_name "/x/x.compute_60.cudafe1.stub.c" --gen_device_file_name "/x/x.compute_60.cudafe1.gpu"  "/x/x.compute_60.cpp1.ii" -o "/x/x.compute_60.ptx"
#$ ptxas -arch=sm_60 -m64  "/x/x.compute_60.ptx"  -o "/x/x.compute_60.cubin" 
#$ gcc -E ... "x.cu" -o "/x/x.compute_70.cpp1.ii" 
#$ "$CICC_PATH/cicc" ... --module_id_file_name "/x/x.module_id" --gen_c_file_name "/x/x.compute_70.cudafe1.c" --stub_file_name "/x/x.compute_70.cudafe1.stub.c" --gen_device_file_name "/x/x.compute_70.cudafe1.gpu"  "/x/x.compute_70.cpp1.ii" -o "/x/x.compute_70.ptx"
#$ ptxas -arch=sm_70 -m64  "/x/x.compute_70.ptx"  -o "/x/x.compute_70.sm_70.cubin" 
```

From the above, we observe that:
* `.cpp1.ii`, `.cudafe1.c`, `.cudafe1.stub.c`, `.cudafe1.gpu`, and `.ptx` files are either:
  * `x.<suffix>`
  * `x.compute_XX.<suffix>`
* `.cubin` files are either:
  * `x.cubin`
  * `x.compute_XX.cubin`
  * `x.compute_XX.sm_XX.cubin`
* without `-c`, the `cicc` command includes `--gen_module_id_file`
* with `-c`, the `cicc` command omits `--gen_module_id_file`

Now that we're hashing all the `cudafe++`, `cicc`, and `ptxas` arguments, these differences yield cache misses when they should be hits. So for simplicity, I updated the renaming logic to rename to the longest form of each (i.e. `x.compute_XX.ptx`, `x.compute_XX.sm_XX.cubin`), and always add the `--gen_module_id_file` flag to `cicc` invocations.

trxcllnt commented:
Looks like integration tests are failing due to https://github.blog/changelog/2025-03-20-notification-of-upcoming-breaking-changes-in-github-actions/#decommissioned-cache-service-brownouts

trxcllnt commented:
cc: @robertmaynard for review

trxcllnt commented on src/compiler/cicc.rs:
```diff
@@ -329,23 +317,22 @@ pub fn generate_compile_commands(
-    Output(PathBuf),
-    PassThrough(OsString),
-    UnhashedFlag,
+    ExtraOutput(PathBuf),
```
I renamed this from `UnhashedOutput -> ExtraOutput` to reflect that these argument types are now hashed.

trxcllnt commented on src/compiler/gcc.rs:
```diff
@@ -905,10 +906,21 @@ where
-                    Language::C => language = Some("cpp-output".into()),
-                    Language::GenericHeader | Language::CHeader | Language::CxxHeader => {}
-                    _ => language.as_mut()?.push_str("-cpp-output"),
+                if let CCompilerKind::Nvhpc = kind {
+                    // -x=c|cpp|c++|i|cpp-output|asm|assembler|ASM|assembler-with-cpp|none
+                    // Specify the language for any following input files, instead of letting
+                    // the compiler choose based on suffix. Turn off with -x none
+                    match parsed_args.language {
+                        Language::C | Language::Cxx => language = Some("cpp-output".into()),
+                        Language::GenericHeader | Language::CHeader | Language::CxxHeader => {}
+                        _ => *(language.as_mut()?) = "none".into(),
+                    }
+                } else {
+                    match parsed_args.language {
+                        Language::C => language = Some("cpp-output".into()),
+                        Language::GenericHeader | Language::CHeader | Language::CxxHeader => {}
+                        _ => language.as_mut()?.push_str("-cpp-output"),
+                    }
```
This change slipped in when I cherry-picked the changes into this branch. It isn't strictly related to `nvcc`, but it does fix the issue using `nvc++` as a host compiler with sccache-dist.

trxcllnt commented on src/compiler/msvc.rs:
```diff
@@ -1025,11 +1021,8 @@ where
-            status,
-            stdout,
-            stderr: stderr_bytes,
-        })
```
The L974-L1032 changes are also unrelated to the functionality, pulled in from another branch by accident. I'll revert them.

sylvestre commented:
i wish this kind of changes would be done in a separate PR 
I can't either:
* squash this PR into a single commit given some unrelated changes
* merge it as it has some commits like "[revert unrelated changes](https://github.com/mozilla/sccache/pull/2383/commits/caf89552f165a08a63c08b60758381bbf76a89d4)"

Maybe split this PR into several

Note: it is why it takes time to merge your PR, split them into smaller PR would make our life much easier

trxcllnt commented:
@sylvestre I am fine with either squashing or merging. If you'd prefer to squash, are there files you'd like me to revert? If merge, I can rebase out the follow-up commits.

drahnr commented:
> Maybe split this PR into several

would be my personal preference.

sylvestre commented:
same, smaller PR would be ideal :)

trxcllnt commented:
I rebased on `main` and squashed the changes in https://github.com/trxcllnt/sccache/commit/ac44e9a6a66fb2f857a37fdf3f4727714c01cdd9, https://github.com/trxcllnt/sccache/commit/3c08b18caa73f161fb60ef8148c71a6671cf0934, and https://github.com/trxcllnt/sccache/commit/caf89552f165a08a63c08b60758381bbf76a89d4 into a single commit.

I can make a separate PR with the test changes after this one. How does that sound?

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -401,25 +407,45 @@ pub fn generate_compile_commands(
+    // up in the preprocessed output, so using random tmpdir paths leads to
+    // erroneous cache misses.
+    let out_dir = env::temp_dir().join("sccache_nvcc").join({
+        // Combine `hash_key` with the output path in case
```
Can we rename `hash_key` to something less ambiguous? It's temp dir subdir stabilizer, `hash_key` is not encoding that, it also needs docs in the `trait` definition.

trxcllnt replied:
I called it `hash_key` because it's the key produced by the call to `generate_hash_key()` for this `nvcc` compilation. I am fine calling it something else, it just seemed the most descriptive and consistent name.

trxcllnt replied:
Done

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -401,25 +407,45 @@ pub fn generate_compile_commands(
+        m.update(cwd.join(output).as_os_str().as_encoded_bytes());
+        m.finish()
+    });
+    fs::create_dir_all(&out_dir).ok();
```
`let _ = fs::create_dir_all(..)` would be preferable, although I don't understand how it'd be acceptable to fail creating the directory?

trxcllnt replied:
That's fine with me

trxcllnt replied:
Done

trxcllnt commented:
I do plan to update the `system.rs` tests to account for the new numbers, it just takes a few hours in the current state and I've been busy with other things lately.

trxcllnt commented:
I updated the tests in `system.rs` so they're all now passing.

The failing rust 1.75.0 jobs are failing to `cargo install grcov`, but that looks to be happening in other PRs and is unrelated to the changes here. Is there a cargo flag or envvar we can set to allow unstable features when installing `grcov`?

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -743,19 +865,38 @@ where
-                            args.splice(idx..(idx + 1), []);
-                        }
+                // Fix for CTK < 12.0:
+                // Remove `--gen_module_id_file` if cudafe++ already does it
```
I am not sure we want to encode `cudafe++` specific behaviour. I pressume `cudafe++` is just one example, and the fix is adjusting the path.

If this is done though, then any custom logic for output dependencies _might_ depend on that. Could you elaborate if you see any further side effects of this?

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -743,19 +865,38 @@ where
+                // of whether the compilation flag is `-c`, `-ptx`, or `-cubin`
+
+                // e.g. test_a.compute_XX.cpp1.ii
+                let mut nidx = args.len() - 3;
```
Did we check anywhere this holds?

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -743,19 +865,38 @@ where
+                let name = name.split(".cpp1.ii").next().unwrap();
+
+                for (flag, name) in [
+                    (&gen_c_file_name_flag, format!("{name}.cudafe1.c")),
```
Could you elaborate the signficance of the `cudafe1.*` file extension? This smells like very application specifc

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -743,19 +865,38 @@ where
+                // Remove `--gen_module_id_file` if cudafe++ already does it
+                if cudafe_has_gen_module_id_file_flag {
+                    if let Some(idx) = args.iter().position(|x| x == &gen_module_id_file_flag) {
+                        args.splice(idx..idx + 1, []);
```
I am pretty sure there is a better way to the above.

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -1114,104 +1237,262 @@ fn fold_env_vars_or_split_into_exe_and_args(
+                            arg
+                        }
+                    })
+                    .collect::<Vec<_>>()
+                    .iter()
```
nit: do we need the intermediate `collect()`?

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -743,19 +865,38 @@ where
+                let mut nidx = args.len() - 3;
+                let name = args[nidx].clone();
+                // test_a.compute_XX.cpp1.ii -> test_a.compute_XX
+                let name = name.split(".cpp1.ii").next().unwrap();
```
Is that intended how it's supposed to work?

trxcllnt replied:
Yes, this is intended. This block is part of ensuring consistent `cicc` arguments, regardless of which `-gencode` flags are or aren't passed.

The input file name is something like `foo.compute75.cpp1.ii`, so the three flags we splice in are:
```shell
--gen_c_file_name foo.compute75.cudafe1.c \
--stub_file_name foo.compute75.cudafe1.stub.c \
--gen_device_file_name foo.compute75.cudafe1.gpu
```

So the logic here extracts the `foo.compute75` component from the input file name and uses it as the prefix for the file name for each of these three flags.

trxcllnt commented on src/compiler/nvcc.rs:
```diff
@@ -743,19 +865,38 @@ where
+                // Fix for CTK < 12.0:
+                // Remove `--gen_module_id_file` if cudafe++ already does it
+                if cudafe_has_gen_module_id_file_flag {
+                    if let Some(idx) = args.iter().position(|x| x == &gen_module_id_file_flag) {
```
Done

trxcllnt commented on src/compiler/nvcc.rs:
```diff
@@ -671,50 +723,123 @@ where
+        if let Some("cudafe++") = exe.file_stem().and_then(|s| s.to_str()) {
+            // Fix for CTK < 12.0:
+            // Add `--gen_module_id_file` if the cudafe++ args include `--module_id_file_name`
+            if let Some(idx) = args.iter().position(|x| x == &module_id_file_name_flag) {
```
Done

drahnr commented on src/compiler/nvcc.rs:
```diff
@@ -516,58 +544,82 @@ impl CompileCommandImpl for NvccCompileCommand {
+                                "[{}]: maybe_keep_temps_then_clean move {src:?} -> {dst:?}",
+                                output_path.display()
+                            );
+                            fs::rename(src, dst)?;
```
Can we ensure it's the same filesystem? Or have a fallback? Otherwise `rename` will fail.

trxcllnt replied:
How would we check if `src` and `dst` are on the same file system? Should I use `fs::copy()` and `fs::remove()` instead of `fs::rename()`?

trxcllnt commented:
Not sure what's going on with Windows2022, but it looks like MSVC's clang version was updated in the last few days? I can try restricting those jobs to CUDA 12.5 instead of 12.8.

drahnr reviewed (approved):
@sylvestre LGTM, you might want to give it a quick pass
